---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, echo = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "README-"
)
```

<h2 style = "text-align: center">genBART</h2>

#### Overview
The goal of `genBART` is to streamline the process of statistical analysis in 
high dimensional studies (e.g. RNAseq and microarray) to a final and 
comprehensive results object that can be uploaded into the BART (Biostatistical 
Analysis Reporting Tool) shiny app. The app provides users with tools to 
interactively visualize and efficiently sift through large amounts of data and 
results. 

The `genBART` package accomplishes its goal by providing a set of functions 
that modularize every step of the analysis workflow: 

1. Summary stats and quality control
2. Unsupervised analysis (heat maps and cluster analysis)
3. DGE (differential gene expression)
4. Gene set analysis
5. Correlation analysis

Though one strength of BART is its ability to report the entire analysis
workflow in one session, it is often useful to use BART before all of the 
analysis is completed (e.g. viewing heat maps before any statistical analysis is 
done). `genBART` makes this possible by allowing users to easily update and/or 
add to existing BART result objects. Conveniently, BART will only populate with 
tools based on the information contained in the object generated by `genBART`.

#### Installation

You can install genBART from github with:

```{r gh-installation, eval = FALSE}
# install.packages("devtools")
devtools::install_github("jcardenas14/genBART")
```

#### Example

I demonstrate a simple example below of how BART can be used for unsupervised 
analysis before any statistical analysis is run. The data used for this example
is available in `genBART` and is taken from a longitudinal microarray experiment
monitoring the gene expression changes in cynomolgus macaques infected with *M.
tuberculosis* (Skinner et al.). To speed up the hierarchical clustering step,
I randomly selected a subset (1000) of the probes for the example. Please see 
the [genBART vignette](vignettes/bart-vignette.rmd) for a full analysis workflow
walk-through from the same microarray study.

``` r
## Call data
library(genBART)
data(tb.expr)
data(tb.design)

# declare design information
des.info <- desInfo(y = tb.expr, design = tb.design, data_type = "micro", 
                    columnname = "columnname", long = TRUE, 
                    patient_id = "monkey_id",baseline_var = "mytimepoint", 
                    baseline_val = 0, time_var = "mytimepoint", 
                    responder_var = "clinical_status",summary_var = "numbChip", 
                    sample_id = "sample_group", project_name = "TB")
                    
# generate dendrograms                    
dendros <- genDendrograms(des.info)

# create BART result object
genFile(design_info = list(des.info), dendros = dendros)
```