---
title: "genBART package Vignette"
author: "Jacob Cardenas"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{genBART package Vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
```{r setup, echo = FALSE, message = FALSE}
knitr::opts_chunk$set(collapse = TRUE, comment = "#>")
library(genBART)
library(limma)
```

BART is a user friendly, point and click, R shiny application that was developed as a biostatistical reporting tool for a broad range of high throughput experiments such as RNA sequencing, Microarray, Flow Cytometry, and Metabolomics. BART provides users with the ability to interactively view and download (all tables and figures generated by the app) the entire analysis workflow:

- Design file and summary statistics 
- Quality control metrics 
- Unsupervised analysis (heatmaps and clustering)
- Differential gene expression (easily sortable genelists, venn diagrams)
- Gene set analysis 
- Correlation analysis

To make the tools that BART offers more readily available, we introduce the genBART package, a series of functions that transforms R objects generated from modeling packages limma, DESeq2, and edgeR into files that can be uploaded into BART. In the following sections, we illustrate how to use the functions available in genBART by walking through the analysis of a longitudinal microarray study tracking gene expression changes in cynomolgus macaques infected with *M. tuberculosis* [(Skinner et al.)](http://www.jimmunol.org/content/197/12/4817.long). In addition to microarray data, the study also has flow data that will be incorporated in to demonstrate the flexibility of the BART app.

<h2 style = "text-align: center">Summary of genBART package</h2>

| Function            | Purpose                                                |
|:--------------------|:------------------------------------------------------ |
| `desInfo()`         | Match design and expression. Define elements of design.|    
| `genDendrograms()`  | normalize and cluster expression.                      |
| `genModules()`      | Generate module (gene set) maps for plotting.          |
| `genModelResults()` | Process and format modeling results.                   |
| `qBart()|rBart()`   | Run gene set analysis using QuSAGE or ROAST algorthims.|
| `crossCorr()`       | Run pairwise correlations between two data sets.       |
| `genFile()`         | Generate and save BART ready .rda file.                |
| `updateFile()`      | Update/add to existing BART file.                      |
| `runBart()`         | Run BART shiny app.                                    |

Table: Functions in the genBART package ordered to demonstrate a typical analysis workflow.

The genBART package contains 10 functions, each of which is listed in the table above. Greater detail on each of these functions and their parameters is available via the help commands in R. Below is a summary of how they are used and integrated into a typical workflow:

1. `desInfo` - Check that the samples in expression match the samples in design. Define elements of experimental design/sample information. This initial step is important as the information provided here is used to determine the various ways the data is normalized, annotate the heat maps produced in BART, and generate module maps for plotting.
2. `genDendrograms` - Use the output from `desInfo` to perform various normalizations (e.g. to healthy controls, baseline samples, etc.) and hiearchical clustering of the genes.
3. `genModules` - Use the output from `desInfo` to generate module maps that show, for each sample, the percentage of probes within a module that are above or below a certain threshold (e.g. mean of healthy controls +- 2 sd). 
4. `genModuleResults` - Use result objects produced by limma, DESeq2, or edgeR to process and format for BART.
5. `qBart|rBart` - Given a list of gene sets, run either QuSAGE (by also including `genModuleResults` output as input in `qBart`) or ROAST, a rotation based approach to gene set testing proposed by Wu et al (2010). 
6. `crossCorr` - Calculate pairwise correlations and p-values between the variables in two data frames or matrices, with the option to correlate by a grouping factor (e.g. time in a longitudinal setting). Flexible to correlate data from different platforms (e.g. microarray vs flow).
7. `genFile` - Generate and save a BART ready .rda file based on the output objects from the functions above. It is not necessary to run all of the functions in order to generate a BART file, making genBART flexible to the project at hand and the stage of analysis a user is at. 
8. `updateFile` - Allows the user to easily update and/or add on to an exisiting BART file without having to re-run all of the functions. 
9. `runBart` - Run the BART shiny app.



### Design Information

Before running any analysis, the first step in generating a BART ready file is to define the elements of the design file and sample information. Keep in mind that not every one of these elements is necessary for the rest of the genBART functions to run, so in practice, much of this section is left to user discretion and the particular project at hand.

```
library(genBART)
library(limma)
```

Below are the first six lines of the design file from the TB study described in the introduction.
```{r}
head(tb.design)
```

For this study, the function call would be:
```{r}
des.info <- desInfo(y = tb.expr, design = tb.design, data_type = "micro", 
                    columnname = "columnname", long = TRUE, patient_id = "monkey_id", 
                    baseline_var = "timepoint", baseline_val = 0, control_var = NULL,
                    control_val = NULL, time_var = "timepoint", summary_var = NULL, 
                    responder_var = "clinical_status", sample_id = "sample_id", 
                    project_name = "TB")
```
Please refer to the R help for `desInfo` for more details on function parameters.

### Clustering (Generate Dendrograms)

Next we generating dendrograms through hierarchical clustering. This step is necessary to produce the heat maps that can be viewed and downloaded in BART. This can be done through the `genDendrograms` function, which takes as input the output from `desInfo`. 

```{r}
dendros <- genDendrograms(design_info = des.info)
```
From the output we can see that the data was normalized in three different ways, and this is determined by the study design and the information given to `desInfo`. For example, if the study included healthy controls then additional normalizations to healthies would have been included. 

### Generate Module Maps

By default, BART automatically generates module maps using a set of 260 coregulated Baylor modules. However, in addition to the default, the user has the option to provide his own gene sets.


### Sample limma Analysis

Now we demonstrate a quick walkthrough of a limma analysis. The example we are using comes from a study in which 38 healthy macaques were infected with *M. tuberculosis*. Data measurements were taken at baseline (before infection) and at 11 additional unequally spaced timepoints up to 6 months post infection. Each macaque was identified as developing symptomatic (active) TB or asymptomatic (latent) infection. For the purposes of illustration and time, a subset of the data was chosen for analysis. We randomly selected 1000 probes, 15 of the 38 macaques, and 5 time points (days 0, 20, 30, 42, and 56). For more details about modeling in limma, refer to the limma User's Guide.

Create a grouping factor by combining clinical status and time point. 
```{r}
tb.design$Group <- paste(tb.design$clinical_status, tb.design$timepoint, sep = "")
grp <- factor(tb.design$Group)

```
Create a design matrix that includes separate coefficients for each level within the grouping factor so that the desired differences can be extracted as contrasts.
```{r}
design2 <- model.matrix(~0+grp)
colnames(design2) <- levels(grp)
```
Since there are repeated measurements on each monkey, we treat monkey_id as a random effect and estimate the correlation between measurements on the same subject. In limma, this is done using duplicateCorrelation.
```{r}
dupcor <- duplicateCorrelation(tb.expr, design2, block = tb.design$monkey_id)
```
Fit model and extract contrasts of interest.
```{r}
fit <- lmFit(tb.expr, design2, block = tb.design$monkey_id, correlation = dupcor$consensus.correlation)
contrasts <- makeContrasts(A_20vsPre = Active20-Active0,A_30vsPre = Active30-Active0,
                           A_42vsPre = Active42-Active0,A_56vsPre = Active56-Active0,
                           L_20vsPre = Latent20-Latent0,L_30vsPre = Latent30-Latent0,
                           L_42vsPre = Latent42-Latent0,L_56vsPre = Latent56-Latent0,
                           levels=design2)

fit2 <- contrasts.fit(fit, contrasts)
fit2 <- eBayes(fit2, trend = FALSE)
```

### Generate Model Results File
The next step in the process is to generate a model results file that is formatted specifically for BART. This can be done by using the `genModelResults` function.

```{r}
mod_results <- genModelResults(design_info = des.info, object = fit2, lm_Fit = fit, method = "limma")
```

### Gene Set Analysis

Next we run gene set analysis by incorporating functions available in the qusage package. The function `geneSetTesting` simplifies the process by requiring only two input parameters.

1. ```object``` - The model results object generated from `genModelResults`.
2. ```gene_sets``` - A list of gene sets.

```
gene.set.test <- geneSetTesting(object = mod_results, gene_sets = genesets)
```

### Correlations
In addition to providing tools for the standard analysis workflow, BART also offers a tool that allows users to visualize and sort through potentially hundreds of thousands pairwise correlations between clinical outcomes. The `crossCorr` function takes two matrices of outcomes and outputs a long format file of all their pairwise correlations. The input parameters are described below.

1. ```x``` - A numeric matrix or dataframe in which the columns are variables. Optional grouping column (i.e. time in longitudinal setting).
2. ```y``` - A second matrix or dataframe with the same number of rows as x
3. ```by```- Defualt is set to NA. If grouping column in x exists, set `by` equal to its columnname.
4. ```method``` - Default is set to "pearson". The alternatives are "spearman" and "kendall".
5. ```description``` - Description of what's being correlated. Default is "X vs Y". Can be any string.
6. ```xvar``` - Description of variables in x matrix. Default is "X". Can be any string.
7. ```yvar``` - Description of variables in y matrix. Default is "Y". Can be any string.
8. ```order.by``` - Order output file by p-value (p) or correlation value (r). Defualt is set to "p" and alternative is "r".
9. ```decreasing``` - Default is TRUE. If FALSE, then the ordering is from smallest to largest.

We walk through an example in which we wish to correlate Baylor modules with probes from the flu dataset that we have been working with. Since module scores are calculated as percentages up or down relative to the mean of the baseline samples, we subtract the mean of the baseline samples from the probes before correlating. For the purposes of this tutorial, we randomly selected 500 probes to correlate with 259 Baylor modules (one of them had zero standard deviation so we threw it out). Since this is a longitudinal setting, we correlate by time.

```
# Add grouping column to module matrix. We don't include baseline since it has been subtracted out.

mods$Hours <- design$Hours[-which(design$Hours == 0)]
corrs <- crossCorr(x = mods, y = exp, by = "Hours", description = "Modules vs Probes", xvar = "Modules", yvar = "Probes")

```

### Generate BART file

The last step in the process is to generate the final file that can be uploaded into BART. This is done through the function `genBartFiles`, which takes as its arguments the objects generated from the previous functions. Conveniently, not every object is required to generate a BART file. For example, if for some reason gene set analysis was not run, the `designInfo`, `genModelResults`, and `genDendrograms` objects can still be run through `genBartFiles`. The BART app will only show the results that are uploaded in.

```
genFile(y = tb.expr, design.info = des.info, modelResults = mod_results, dendrograms = dendros, geneSetTest = gene.set.test, corr = corrs, Illumina = FALSE)
```



